### PR TITLE
Add recipe for ChefFun 1.0.1 library

### PR DESCRIPTION
**Pull request by maxpagani**
Specify library name and version:  **chef-fun/1.0.1**

Chef-fun is a self-contained part of a larger library (Chef). This subset is aimed at enabling a functional programming style for C++ development. The library is OSS (Apache2.0 license).

I am the author and I like to ease the adoption of this library and also use it as a part of a larger codebase I'm working on in my daily job.
---

- [X] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [X] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [X] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Review by RubenRBS — APPROVED** [review]
Looks good, thank you for taking the time to contribute this recipe and for your patience while we migrated our infrastructure to Conan v2 :)

File: recipes/chef-fun/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    @property
+    def _compilers_minimum_version(self):
+        return {
+            "gcc": "9.4.0",
+        }
+
+    # Use the export_sources(self) method instead of the exports_sources attribute.
+    # This allows finer grain exportation of patches per version
+    def export_sources(self):
+        export_conandata_patches(self)
```
Comment (prince-chrismc) [inline recipes/chef-fun/all/conanfile.py:41]:
```suggestion
```

File: recipes/chef-fun/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+    def layout(self):
+        # src_folder must use the same source folder name than the project
```
Comment (prince-chrismc) [inline recipes/chef-fun/all/conanfile.py:44]:
```suggestion
```

File: recipes/chef-fun/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+#        self.requires("dependency/0.8.1", transitive_headers=True)
+
+    def source(self):
+        # Download source package and extract to source folder
+        get(self, **self.conan_data["sources"][self.version], strip_root=True)
+
+    # Not mandatory when there is no patch, but will suppress warning message about missing build() method
+    def build(self):
+        # The attribute no_copy_source should not be used when applying patches in build
+        apply_conandata_patches(self)
```
Comment (prince-chrismc) [inline recipes/chef-fun/all/conanfile.py:73]:
```suggestion
            )

#    def source(self):
        get(self, **self.conan_data["sources"][self.version], strip_root=True)
```

File: recipes/chef-fun/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+        ## Add m, pthread and dl if needed in Linux/FreeBSD
+        #if self.settings.os in ["Linux", "FreeBSD"]:
+        #    self.cpp_info.system_libs.extend(["dl", "m", "pthread"])
+
+        # TODO: to remove in conan v2 once cmake_find_package_* generators removed
+        self.cpp_info.names["cmake_find_package"] = "chef-fun"
+        self.cpp_info.names["cmake_find_package_multi"] = "chef-fun"
+        self.cpp_info.filenames["cmake_find_package"] = "chef-fun"
+        self.cpp_info.filenames["cmake_find_package_multi"] = "chef-fun"
```
Comment (prince-chrismc) [inline recipes/chef-fun/all/conanfile.py:117]: https://gitlab.com/libchef/chef-fun/-/blob/main/CMakeLists.txt

```suggestion
        self.cpp_info.libdirs = []
```

**Review by prince-chrismc — COMMENTED** [review]
I am glad to see you took advantage of the templates but here's a lot of extras which are not required for this recipe :)

I dropped some suggestions, you'll probably need to clean up the header too

**Review by SSE4 — APPROVED** [review]

**Pull request merged**